content/articles/working-with-java-synchronization/index.md

```diff
@@ -0,0 +1,294 @@
+          }
+        }
+      }).start();
+    }	
+  }
+}
+
+```
+
+```
+The output of the thread is : Thread-0 - 2
+The output of the thread is : Thread-4 - 3
+The output of the thread is : Thread-3 - 4
+The output of the thread is : Thread-2 - 5
+The output of the thread is : Thread-1 - 6
+```
+
+From the results of the threads above, no thread shares a value. Our aim of avoiding a race condition is fulfilled.
+
+### Understanding synchronized methods and synchronized blocks
+To understand the two, let us first look at each separately:
+
+#### Synchronized methods
+These consist of the following properties:
+
+- A synchronized method locks down the whole object. No other thread may access synchronized methods in the object while the method is executing. With static methods, they are locked by their class.
+- Synchronized methods need locks on the current object or, in the case of static methods, the whole class. This is because the lock is obtained when the thread enters and released when the thread quits (naturally or by throwing an exception).
+- Synchronized methods keep the lock for the method scope.
+- A synchronized static function prevents instances from being changed. 
+
+#### Synchronized blocks
+The synchronized block may be used to perform synchronization on any specified resource of the method.
+
+Suppose a method has 100 lines of code but only 10 lines of code containing important logic. I.e. the object's state may be modified by these lines. Only 10 lines of code are needed to guarantee that other threads may continue to run inside the same procedure without interruption. Thus we don't need to synchronize these lines.
+
+Let us look at an example program that uses synchronized blocks.
+
+Save it as `TestSynchronizedBlock1.java`
+
+```java
+    class Table  
+    {      
+     void printTable(int n){    
+       synchronized(this){//synchronized block    
+         for(int i=1;i<=5;i++){    
+          System.out.println(n*i);    
+          try{    
+           Thread.sleep(400);    
+          }catch(Exception e){System.out.println(e);}    
+         }    
+       }    
+     }//end of the method    
+    }    
+        
+    class MyThread1 extends Thread{    
+    Table t;    
+    MyThread1(Table t){    
+    this.t=t;    
+    }    
+    public void run(){    
+    t.printTable(5);    
+    }    
+        
+    }    
+    class MyThread2 extends Thread{    
+    Table t;    
+    MyThread2(Table t){    
+    this.t=t;    
+    }    
+    public void run(){    
+    t.printTable(100);    
+    }    
+    }    
+        
+    public class TestSynchronizedBlock1{    
+    public static void main(String args[]){    
+    Table obj = new Table();//only one object    
+    MyThread1 t1=new MyThread1(obj);    
+    MyThread2 t2=new MyThread2(obj);    
+    t1.start();    
+    t2.start();    
+    }    
+    }    
+```

```

Comment:
There seems to be a race condition issue with this code. Occasionally, I noticed it would output:

```
100
200
300
400
500
5
10
15
20
25
```
Please fix it

Reply:
![Capture](https://user-images.githubusercontent.com/63324129/161399350-732ce91e-e89b-4630-8930-b5c9c7c4f4ba.PNG)